### PR TITLE
Non backtracking

### PR DESCRIPTION
**Pull request by jpfairbanks**
This PR is updates PR #196 to add the non_backtracking random walks. This PR includes @CarloLucibello 's code plus one more set of tests which are more general than the tests proposed in PR #196. Thanks @CarloLucibello for writing this and being so amenable to changes.

### CODE REVIEW

File: src/traversals/randomwalks.jl [hunk 1]
```diff
@@ -17,6 +17,68 @@ function randomwalk(g::SimpleGraph, s::Integer, niter::Integer)
...
+    length(nbrs) == 0 && return visited[1:i-1]
+    prev = currs
+    currs = rand(nbrs)
+
+    while i <= niter
+        push!(visited, currs)
+        i += 1
+        nbrs = out_neighbors(g,currs)
+        length(nbrs) == 1 && break
+        idnext = rand(1:length(nbrs)-1)
```
Comment (sbromberger) [inline src/traversals/randomwalks.jl]: I like the rand(nbrs) in DiGraph. Why are we doing this differently here, and can we change the other functions as well?

File: src/traversals/randomwalks.jl [hunk 2]
```diff
@@ -17,6 +17,68 @@ function randomwalk(g::SimpleGraph, s::Integer, niter::Integer)
...
+
+    while i <= niter
+        push!(visited, currs)
+        i += 1
+        nbrs = out_neighbors(g,currs)
+        length(nbrs) == 0 && break
+        next = rand(nbrs)
+        if next == prev
+            length(nbrs) == 1 && break
+            idnext = rand(1:length(nbrs)-1)
```
Comment (sbromberger) [inline src/traversals/randomwalks.jl]: I can't remember why I used idnext in my other functions. `next = rand(nbrs)` should work, no?

Comment (jpfairbanks) [inline src/traversals/randomwalks.jl]: we need to sample from `1:deg-1` in order to make the probabilities work out. If we sample from `rand(nbrs)` directly, we would oversample `nbrs[end]`

Comment (sbromberger) [inline src/traversals/randomwalks.jl]: gah. of course. OK. I think this is ready to merge.

File: src/traversals/randomwalks.jl [hunk 3]
```diff
@@ -17,6 +17,68 @@ function randomwalk(g::SimpleGraph, s::Integer, niter::Integer)
...
+    nbrs = out_neighbors(g,currs)
+    length(nbrs) == 0 && return visited[1:i-1]
+    prev = currs
+    currs = rand(nbrs)
+
+    while i <= niter
+        push!(visited, currs)
+        i += 1
+        nbrs = out_neighbors(g,currs)
+        length(nbrs) == 1 && break
```
Comment (sbromberger) [inline src/traversals/randomwalks.jl]: what happens if there are no neighbors? (Shouldn't this be <= 1?)

```
julia> g = PathDiGraph(5)
{5, 4} directed graph

julia> out_neighbors(g,5)
0-element Array{Int64,1}
```

**Comment by jpfairbanks** [comment]
Tests pass: :ship: it.

**Comment by sbromberger** [comment]
whoa - careful here. You merged into `v4`, not `master`....

File: src/traversals/randomwalks.jl (hunk 3 above)
Comment (jpfairbanks) [inline src/traversals/randomwalks.jl]: yes it should be.

**Comment by jpfairbanks** [comment]
That was where PR #196 was. When will we be merging v4 into master?

**Comment by sbromberger** [comment]
I merged it on Wednesday. No worries though - I managed to pull this back into master. From here on we should not be making PRs against any branches but 0.3.6 (only for urgent fixes) and master.

**Pull request merged**